### DISCUSSION TITLE
how to do data binding?

### DISCUSSION THREAD

**Discussion by shi-yan** [gyscos/cursive#632] [Q&A]
how can I do data binding using cursive, for example if I have 

struct Content {
content1: String,
content2: String,
content3: String
}

 and create one text area for each content. I want the data and the ui be in sync?

textarea can't hold a mutable reference to a piece of content, and in textarea's callback function, we don't allow to mutate values from outside (it's not FnMut).

I don't know what to do here?

In the official tutorial_3, the above situation is not covered (https://github.com/gyscos/cursive/blob/main/doc/tutorial_3.md)

as the tutorial only shows how to mutate the cursive ui struct itself.

however, in a model-view-controller setup, the data is not held directly by the ui. In this case, I don't know how to update the data given the callback function isn't FnMut?

**Comment by gyscos** [comment]
> textarea can't hold a mutable reference to a piece of content, and in textarea's callback function, we don't allow to mutate values from outside (it's not FnMut).

Most callbacks are not `FnMut` because they might be re-entrant: a callback can trigger events on the given `&mut Cursive`, potentially triggering the same callback. That's not possible with `FnMut`.

If you don't care about this, there are tricks using interior mutability. A few macros are defined in cursive to make this convenient to use: for example [`immut1!`](https://docs.rs/cursive/latest/cursive/macro.immut1.html), [`immut2!`](https://docs.rs/cursive/latest/cursive/macro.immut2.html) and [`immut3!`](https://docs.rs/cursive/latest/cursive/macro.immut3.html). As the doc says:
> If this function tries to call itself recursively (for example by triggering an event in `Cursive`), the second call will be a no-op. Enabling recursive calls would break the `FnMut` contract.

In a different direction, it's also possible to write custom views that are most friendly to MVC configurations. It may include storing the content in a shared data structure. `TextView` starts something along this line, though it's only very early (and without other views doing the same, it doesn't let you easily connect views together).
